tools/gen_ut_cmakelists.py

```diff
@@ -161,146 +162,330 @@ def process_name(name, curdir):
+        generate new CmakeLists.txt files.
+    ====================================================================================
     """
-
-    name, os_, archs, timeout, run_type, launcher, dist_ut_port, run_serial, envs, conditions = line.strip(
-    ).split(",")
-
-    # name == "name" means the line being parsed is the header of the table
-    # we should skip this line and return empty here.
-    if name == "name":
-        return ""
-    name = process_name(name, curdir)
-
-    envs = process_envs(envs)
-    conditions = process_conditions(conditions)
-    archs = proccess_archs(archs)
-    os_ = process_os(os_)
-    run_serial = process_run_serial(run_serial)
-    run_type = process_run_type(run_type)
-
-    cmd = ""
-
-    for c in conditions:
-        cmd += f"if ({c})\n"
-
-    time_out_str = f'TIMEOUT "{timeout}"' if len(timeout.strip()) > 0 else ''
-    if launcher[-3:] == ".sh":
-        dist_ut_port = process_dist_ut_port(2)
-        cmd += f'''if({archs} AND {os_})
-    bash_test_modules(
-    {name}
-    START_BASH
-    {launcher}
-    LABELS
-    "RUN_TYPE={run_type}"
-    ENVS
-    "PADDLE_DIST_UT_PORT={dist_ut_port};{envs}")
-    set_tests_properties({name} PROPERTIES  {time_out_str} RUN_SERIAL {run_serial})
-endif()
-'''
-    else:
-        cmd += f'''if({archs} AND {os_})
-    py_test_modules(
-    {name}
-    MODULES
-    {name}
-    ENVS
-    "{envs}")
-    set_tests_properties({name} PROPERTIES {time_out_str} RUN_SERIAL {run_serial})
-endif()
-'''
-    for _ in conditions:
-        cmd += f"endif()\n"
-    return cmd
-
-
-PROCESSED_DIR = set()
-
-
-def gen_cmakelists(current_work_dir):
-    print("procfessing dir:", current_work_dir)
-    if current_work_dir == "":
-        current_work_dir = "."
-
-    contents = os.listdir(current_work_dir)
-    sub_dirs = []
-    for c in contents:
-        c_path = os.path.join(current_work_dir, c)
-        if c_path in PROCESSED_DIR:
-            return
-        if os.path.isdir(c_path):
-            PROCESSED_DIR.add(c_path)
+            assert len(self.no_cmake_dirs) == 0, err_msg
+
+
+class CMakeGenerator():
+
+    def __init__(self, current_dirs):
+        self.processed_dirs = set()
+        self.port_manager = DistUTPortManager()
+        self.current_dirs = current_dirs
+
+    def prepare_dist_ut_port(self):
+        for c in self._find_root_dirs():
+            self.port_manager.parse_assigned_dist_ut_ports(
+                c, ignores=args.ignore_cmake_dirs, depth=0)
+
+    def parse_csvs(self):
+        for c in self.current_dirs:
+            c = os.path.abspath(c)
+            self._gen_cmakelists(c)
+
+    def _find_root_dirs(self):
+        root_dirs = []
+        # for each current directory, find its highest ancient directory (at least itself)
+        # which includes CMakeLists.txt or testslist.csv.txt in the filesys tree
+        for c in self.current_dirs:
+            c = os.path.abspath(c)
+            while True:
+                ppath = os.path.dirname(c)
```

Comment:
https://stackoverflow.com/questions/2860153/how-do-i-get-the-parent-directory-in-python

Reply:
https://www.geeksforgeeks.org/get-parent-of-current-directory-using-python/